### PR TITLE
Fairspc 108 metadata page count

### PR DESCRIPTION
**Pull request by tgreenwood**

### CODE REVIEW

File: projects/saturn/src/main/java/io/fairspace/saturn/config/ViewsConfig.java [hunk 1]
```diff
@@ -1,18 +1,30 @@
...
+import java.util.stream.Collectors;
 
 import com.fasterxml.jackson.annotation.*;
 import com.fasterxml.jackson.databind.*;
 import com.fasterxml.jackson.dataformat.yaml.*;
 import jakarta.validation.constraints.*;
 
 public class ViewsConfig {
+
+    private Map<String, View> viewConfig;
```
Comment (ewelinajozwik) [inline projects/saturn/src/main/java/io/fairspace/saturn/config/ViewsConfig.java]: Could you rename this to something like `viewConfigMap` or `nameToViewConfigMap`? 
It is confusing that `getViewConfig` gives you `View`, but `viewConfig` is a map of names to `Views`.

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java [hunk 2]
```diff
@@ -169,14 +195,10 @@ private ValueDTO toValueDTO(RDFNode node) {
...
-    private Query getQuery(CountRequest request) {
+    private Query getQuery(CountRequest request, boolean isCount) {
         var view = getView(request.getView());
 
-        var builder = new StringBuilder("PREFIX fs: <")
-                .append(FS.NS)
-                .append(">\n\nSELECT ?")
-                .append(view.name)
-                .append("\nWHERE {\n");
+        var builder = new StringBuilder().append("SELECT ").append("%s").append("\nWHERE {\n");
```
Comment (ewelinajozwik) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java]: why this change? Why not putting the view name already here, instead of the '%s'? And why adding the prefix later?

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java [hunk 3]
```diff
@@ -376,23 +419,25 @@ private static boolean convertBooleanValue(String value) {
...
+                    return new CountDTO(-1, false);
                 }
-                return new CountDTO(count, false);
-            } catch (QueryCancelledException e) {
-                return new CountDTO(count, true);
-            }
-        });
+            });
+        } catch (QueryCancelledException e) {
+            return new CountDTO(-1, true);
```
Comment (ewelinajozwik) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java]: SparqlQueryService and JdbcQueryService return different values in this case now. Previously both returned 0, now Sparql returns -1, Jdbc 0.

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewConfigDto.java [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package io.fairspace.saturn.services.views;
+
+import lombok.Value;
+
+import io.fairspace.saturn.config.ViewsConfig;
+
+@Value
+public class ViewConfigDto {
```
Comment (ewelinajozwik) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewConfigDto.java]: This class is not used anywhere...

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java
Comment (ewelinajozwik) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java:1]: Please replace line 174 with the new function `viewsConfig.getViewConfig(j.view)`

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java [hunk 5]
```diff
@@ -189,7 +189,11 @@ protected List<ViewDTO> fetchViews() {
                             columns.add(new ColumnDTO(joinView.name + "_" + c.name, c.title, c.type, j.displayIndex));
                         }
                     }
-                    return new ViewDTO(v.name, v.title, columns);
+                    Long maxDisplayCount = viewsConfig
+                            .getViewConfig(v.name)
+                            .map(t -> t.maxDisplayCount)
+                            .orElse(null);
+                    return new ViewDTO(v.name, v.title, columns, maxDisplayCount);
```
Comment (ewelinajozwik) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java]: why retrieving maxDisplayCount again, if you have it available as `v.maxDisplayCount`?

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java (hunk 2 above)
Comment (ewelinajozwik) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java]: ok, I understand now, but the query building logic is now hard to follow...

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java (hunk 3 above)
Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java]: Agree, now it is different. In turn, I am not sure that 0 is the right value in case of something went wrong as potentially it is valid count value while -1 is not valid highlighting that something went wrong. Will put 0 for now

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java (hunk 5 above)
Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java]: we extract the value if none null ViewConfig found by name

File: projects/saturn/src/main/java/io/fairspace/saturn/config/ViewsConfig.java (hunk 1 above)
Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/config/ViewsConfig.java]: agree

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewConfigDto.java (hunk 4 above)
Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewConfigDto.java]: good catch

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java (hunk 5 above)
Comment (ewelinajozwik) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java]: but you are already iterating through views and you can have that value here from `v.maxDisplayCount`, can't you?

Comment (ewelinajozwik) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java]:
```suggestion
                    return new ViewDTO(v.name, v.title, columns, v.maxDisplayCount);
```

Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java]: haha, indeed, facepalm

Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java]: applied

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewConfigDto.java (hunk 4 above)
Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewConfigDto.java]: removed

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java (hunk 3 above)
Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java]: put 0

File: projects/saturn/src/main/java/io/fairspace/saturn/config/ViewsConfig.java (hunk 1 above)
Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/config/ViewsConfig.java]: renamed

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java (hunk 2 above)
Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java]: agree, slightly improved, but the prefix to be added in the end anywhere due to potential query wrapping for count query. What do you think now?

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java [hunk 6]
```diff
@@ -189,7 +186,7 @@ protected List<ViewDTO> fetchViews() {
                             columns.add(new ColumnDTO(joinView.name + "_" + c.name, c.title, c.type, j.displayIndex));
                         }
                     }
-                    return new ViewDTO(v.name, v.title, columns);
```
Comment (ewelinajozwik) [inline projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java:192]: Could you update the documentation (updates to the API endpoint)?

File: projects/saturn/src/main/java/io/fairspace/saturn/config/ViewsConfig.java [hunk 7]
```diff
@@ -68,6 +80,12 @@ public static class View {
          * The name of the items that appear as rows.
          */
         public String itemName;
+        /**
+         * The max count value to be requested for a view total count.
+         * If total count is greater than this max value, the total count value will look like 'more than 1000' on FE.
+         * This is to prevent performance issues when the total count is too large.
+         */
+        public Long maxDisplayCount;
```
Comment (ewelinajozwik) [inline projects/saturn/src/main/java/io/fairspace/saturn/config/ViewsConfig.java:88]: Maybe with @Positive annotation? Then it needs to be > 0 or null.
```suggestion
        @Positive
        public Long maxDisplayCount;
```
We don't validate enough the values of this config, there is a ticket to improve it, but we can already start making it better. I guess the negative value of limit in Postgres doesn't make sense

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java (hunk 2 above)
Comment (ewelinajozwik) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/SparqlQueryService.java]: sure, looks good!

File: projects/saturn/src/main/java/io/fairspace/saturn/config/ViewsConfig.java (hunk 7 above)
Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/config/ViewsConfig.java:88]: I would do it in the scope of this ticket https://thehyve.atlassian.net/browse/FAIRSPC-80. Now there are a few annotations like @NotBlank in the properties class, but they don't work, validation is not triggered anywhere. So, I would leave it for now, but I dropped it in the 80th ticket description

File: projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java (hunk 6 above)
Comment (tgreenwood) [reply projects/saturn/src/main/java/io/fairspace/saturn/services/views/ViewService.java:192]: good point, done

File: README.adoc [hunk 8]
```diff
@@ -1239,7 +1239,7 @@ curl -X POST -H 'Content-type: application/json' -H 'Accept: application/json' -
 |===
 3+| ``POST /api/views/count``
 
-3+| Count rows of a view matching request filters.
+3+| Count rows of a view matching request filters. If `maxDisplayCount` configured in the `views.ttl` for a view, then the count for the view is limited by this value if total count exceeds it. Otherwise, the total count is returned.
```
Comment (ewelinajozwik) [inline README.adoc]: `views.yaml`, not `views.ttl`

Comment (tgreenwood) [reply README.adoc]: oops, fixed

**Review by ewelinajozwik — APPROVED** [review]
